src/Microsoft.Diagnostics.Monitoring/DiagnosticServices.cs

```diff
@@ -68,25 +69,28 @@ public async Task<Stream> GetDump(int pid, DumpType mode)
             return new AutoDeleteFileStream(dumpFilePath);
         }
 
-        public async Task<Stream> GetGcDump(int pid, TimeSpan timeout, CancellationToken cancellationToken)
+        public async Task<Stream> GetGcDump(int pid, CancellationToken cancellationToken)
         {
-            using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
-            linkedSource.CancelAfter(timeout);
-
             var graph = new MemoryGraph(50_000);
             await using var processor = new DiagnosticsEventPipeProcessor(_contextConfiguration,
                 PipeMode.GCDump,
                 gcGraph: graph);
 
-            await processor.Process(pid, Timeout.InfiniteTimeSpan, linkedSource.Token);
+            await processor.Process(pid, Timeout.InfiniteTimeSpan, cancellationToken);
 
-            string dumpFilePath = Path.Combine(Path.GetTempPath(), FormattableString.Invariant($"{Guid.NewGuid()}_{pid}"));
-            GCHeapDump.WriteMemoryGraph(graph, dumpFilePath, "dotnet-monitor");
+            var dumper = new GCHeapDump(graph);
+            dumper.CreationTool = "dotnet-monitor";
 
-            return new AutoDeleteFileStream(dumpFilePath);
+            var stream = new MemoryStream();
+            var serializer = new Serializer(stream, dumper, leaveOpen: true);
+            serializer.Close();
+
+            stream.Position = 0;
```

Comment:
I'm not sure how complex the Serializer is (and more importantly if it does rewrites or random access), but can you remove the intermediate memory stream here?. You can use OutputStreamResult in DiagController to push across the response stream directly.

Reply:
It's calling at least one method (get_Length) that is not supported on the BodyWrapperStream. I'm going to leave it as-is.